sled-agent/src/sp.rs

```diff
@@ -0,0 +1,232 @@
+    /// Currently the only "detection" performed is whether `sp_config` is
+    /// `Some(_)`, in which case a simulated SP is started, and a handle to it
+    /// is returned.
+    ///
+    /// A return value of `Ok(None)` means no SP is available.
+    pub async fn detect(
+        sp_config: &Option<GimletConfig>,
+        sled_config: &SledConfig,
+        log: &Logger,
+    ) -> Result<Option<Self>, SpError> {
+        let inner = if let Some(config) = sp_config.as_ref() {
+            let sim_sp = start_simulated_sp(config, sled_config, log).await?;
+            Some(Inner::SimulatedSp(sim_sp))
+        } else {
+            None
+        };
+        Ok(inner.map(|inner| Self { inner }))
+    }
+
+    pub fn manufacturing_public_key(&self) -> Ed25519PublicKey {
+        match &self.inner {
+            Inner::SimulatedSp(sim) => sim.sp.manufacturing_public_key(),
+        }
+    }
+
+    // TODO The error type here leaks that we only currently support simulated
+    // SPs and will need work once we support a real SP.
+    pub fn rot_handle(&self) -> RotManagerHandle<SimRotTransportError> {
+        match &self.inner {
+            Inner::SimulatedSp(sim) => sim.rot_handle.clone(),
+        }
+    }
+
+    pub fn rot_certs(&self) -> Ed25519Certificates {
+        match &self.inner {
+            Inner::SimulatedSp(sim) => sim.rot_certs,
+        }
+    }
+}
+
+#[derive(Clone)]
+enum Inner {
+    SimulatedSp(SimulatedSp),
+}
+
+#[derive(Clone)]
+struct SimulatedSp {
+    sp: Arc<sp_sim::Gimlet>,
+    rot_certs: Ed25519Certificates,
+    rot_handle: RotManagerHandle<SimRotTransportError>,
+}
+
+async fn start_simulated_sp(
+    sp_config: &GimletConfig,
+    sled_config: &SledConfig,
+    log: &Logger,
+) -> Result<SimulatedSp, SpError> {
+    // Is our simulated SP going to bind to addresses (acting like management
+    // network IPs)?
+    if let Some(bind_addrs) = sp_config.common.bind_addrs {
+        // Sanity check that the sim SP config only specifies one IP address; we
+        // can simulate multiple management network ports by using different TCP
+        // ports.
+        let sp_addr = bind_addrs[0].ip();
+        for addr in bind_addrs[1..].iter().copied().chain(
+            sp_config.components.iter().filter_map(|comp| comp.serial_console),
+        ) {
+            if sp_addr != addr.ip() {
+                return Err(SpError::SimulatedSpMultipleIpAddresses(
+                    *sp_addr,
+                    *addr.ip(),
+                ));
+            }
+        }
+
+        // Ensure we have the global zone IP address we need for the SP.
+        let data_link = if let Some(link) = sled_config.data_link.clone() {
+            link
+        } else {
+            Dladm::find_physical()?
+        };
+        Zones::ensure_has_global_zone_v6_address(data_link, *sp_addr, "simsp")
+            .map_err(|err| SpError::EnsureGlobalZoneAddressError {
+                addr: *sp_addr,
+                err,
+            })?;
```

Comment:
As of https://github.com/oxidecomputer/omicron/pull/1066 , I've moved most of the VNIC / address allocation to happen over an etherstub instead of a physical data link. [motivation here](https://github.com/oxidecomputer/opte/issues/127#issuecomment-1104706742).

As a result, most of our addresses are being allocated over these etherstubs; there's an example here: 

https://github.com/oxidecomputer/omicron/blob/813a859638ba72e4cdf0716e41b630fc9d7dcb10/sled-agent/src/bootstrap/agent.rs#L157-L177

Could this `sp_addr` also be allocated on the etherstub link?

Reply:
I expect so; I'll rebase to confirm. For now I haven't been testing the management network side of the simulated SPs at all (and am running a test config without any networking), since we only need the simulated RoT to do sprockets things. This will be revisited as we move back to MGS and start integrating it into the control plane, at which point it will need to access the (simulated) SPs over UDP.

Reply:
Changed in 02c2f578; we not only could change this allocation, but had to: `Zones::ensure_has_global_zone_v6_address()` now takes an `EthernetVnic` as its first arg, so the previous code allocating on the physical link no longer compiled.